### ISSUE TITLE
keepalive config ignored

### ISSUE THREAD

**Issue by jtgrassie** [xmrig/xmrig#2195]
When a pool reports it supports the keepalive extension, the the miner will always send keepalive messages, regardless if the user has specified `keepalive:false` or `keepalive:0` in their miner config.

**Comment by xmrig** [comment]
This is how it is supposed to be: If pool put `keepalive` to extensions list it means pool supports it and wants this feature to be enabled. It's zero configuration same way as pool override `algo` option.

If a pool didn't put it to the extensions list or don't have this list all works the same way since the beginning of time. This is wrong https://github.com/jtgrassie/monero-pool/pull/74 `keepalive` option work as expected. Likely confused with this https://github.com/xmrig/xmrig/blob/master/src/base/net/stratum/Client.h#L152
Thank you.

**Comment by jtgrassie** [comment]
If the client (xmrig) sets `keepalive:false` in their config, I'd expect the client would *not* send keepalives. The fact a pool does or does not support a feature should not dictate whether a client uses that feature.

**Comment by xmrig** [comment]
Behavior of `keepalive:true` not changed just added one extra way to enable it (on pool side).
Thank you.

**Comment by jtgrassie** [comment]
> ...just added one extra way to enable it (on pool side).

I'm sorry, but this is not the case. Here is what happens with and without the patch...

Without #2194:

- If a pool doesn't send the extension=keepalive, no matter what miner config is used, no keepalives will be sent, period. This seems reasonable, although before the extension was added, keepalives would be sent if configured in the miner (commandline flag or config.json).
- If a pool _does_ flag support in the login extension, _there is no way_ for the miner to disable (keepalive:false|0 config/flag is simply ignored). This is the crux of the problem.

With #2194:

- If the pool sends the extension _and_ the miner has keepalive:true|N>0, then keepalives are sent. This seems the intuitively the correct behavior, as the pool is saying whether it supports keepalive, but allows the miner to choose (configure) if they want to send them or not.

**Comment by xmrig** [comment]
Did you check it https://github.com/xmrig/xmrig/blob/master/src/base/net/stratum/Client.h#L152 my quick check shows it works as expected. Maybe I miss some bug that prevent keepalive sent? If `keepalive:true` not work without extension this is bug.

**Comment by jtgrassie** [comment]
> Did you check ... my quick check shows it works as expected. Maybe I miss some bug that prevent keepalive sent? If keepalive:true not work without extension this is bug.

Yes i did check and no keepalive is sent unless the pool sends support via the login extension. That's why moneromoo added the patch to the pool in the first place. It was then after more testing, I discovered the miner cannot disable it (when the pool says it can handle it). Hence the PR here and detail in the [comment](https://github.com/xmrig/xmrig/issues/2195#issuecomment-803268064) above.

**Comment by jtgrassie** [comment]
https://github.com/xmrig/xmrig/blob/master/src/base/net/stratum/Client.h#L152 is the crux of the problem, because if the pool sends support for keepalive extension, the miner cannot switch off sending keepalives (no matter what they configure, keepalives will be sent because the pool says it supports them).

**Comment by xmrig** [comment]
Sorry I can't confirm the bug https://i.imgur.com/W5gS6kn.png don't forget if you must use high diff, every share submit resets keepalive timer.

Purpose of keepalive extension is to override user defined keepalive option, you PR breaks it, so it breaks xmrig-proxy and other pools who rely on this behavior, eg https://hashvault.pro/

**Comment by jtgrassie** [comment]
> Purpose of keepalive extension is to override user defined keepalive option

Well in that case we don't want that extension in the pool. I'll revert the PR.

**Comment by xmrig** [comment]
And please record there was none of breaking changes in ~2019.
Thank you.

**Comment by jtgrassie** [comment]
I never said it was a _breaking_ change in 2019, just that there were changes in behavior - no harm no foul.

And we couldn't see keepalives happening _without_ the login extension being sent from the pool. You've detailed how you intend it to be enforced by the pool, which you have your reasons for, which is my misunderstanding for assuming the miner should be able to enable/disable it (per the config or commandline option).

**Comment by xmrig** [comment]
> And we couldn't see keepalives happening without the login extension being sent from the pool.

But https://i.imgur.com/W5gS6kn.png hint: supportxmr.com doesn't support login extension.

**Comment by jtgrassie** [comment]
Fine, maybe we needed to set a higher diff or something. We both tested and clearly both made the same mistake. The PR for this issue was about the inability of the miner to switch off keepalives if the pool reported via the extension. As I said, if that's how you intend it to work then fine, I've closed the issue and PR. Apologies for implementing in the pool and sending a PR for the miner for what was assumed a bug not a feature.
